src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -302,5 +356,74 @@ std::string ObjectCacheStore::get_cache_file_path(std::string cache_file_name,
   return m_cache_root_dir + cache_file_dir + cache_file_name;
 }
 
+void ObjectCacheStore::handle_throttle_ready(void* ctx, uint64_t flag) {
+  m_io_throttled.store(false);
+}
+
+bool ObjectCacheStore::take_token_from_throttle(uint64_t object_size,
+                                                uint64_t object_num) {
+  bool wait = false;
+
+  if (m_io_throttled.load() == true) {
+    return false;
+  }
+
+  if (!wait && (m_qos_enabled_flag & ROC_QOS_IOPS_THROTTLE)) {
+    wait =
+      m_throttles[ROC_QOS_IOPS_THROTTLE]->get(object_num, this,
+      &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
+  }
+
+  if (!wait && (m_qos_enabled_flag & ROC_QOS_BPS_THROTTLE)) {
+    wait =
+      m_throttles[ROC_QOS_BPS_THROTTLE]->get(object_size, this,
+      &ObjectCacheStore::handle_throttle_ready, nullptr, 0);
+  }
+
+  // TODO : better implement..
+  if (wait) {
+    m_io_throttled.store(true);
+  }
+
+  return !wait;
+}
+
+static std::map<uint64_t, std::string> throttles_flag = {
```

Comment:
```suggestion
static const std::map<uint64_t, std::string> THROTTLE_FLAGS = {
```

Reply:
> 

If I add “const”, I will report an error when I use it to pass parameters.
error: passing ‘const std::map<long unsigned int, std::__cxx11::basic_string<char> >’ as ‘this’ argument discards qualifiers [-fpermissive]
412 |   throttle = new TokenBucketThrottle(m_cct, THROTTLE_FLAGS[flag],

Reply:
Right -- you need to use something like:

```
auto throttle_flags_it = THROTTLE_FLAGS.find(flag);
ceph_assert(throttle_flags_it != THROTTLE_FLAGS.end());
throttle = new TokenBucketThrottle(m_cct, throttle_flags_it->second,
    ...);
```